### ISSUE TITLE
Create multiple population capability in CoLoRe

### ISSUE THREAD

**Issue by slosar** [DarkEnergyScienceCollaboration/2pt_validation#4]
I.e. instead of N(z) and b(z), have N(M), b(M) and p(M,z). The idea is that to mock up "halo mass" effect. At this level no need for stochasticity.

**Comment by damonge** [comment]
I can do this. A simple way would be to discretize over M, such that you tell it the number of populations and give it files for the b(z) and n(z) of each. Would this work?

This should be trivial to implement. Implementing a full color+redshift-space distribution function would take longer (and might need input from others...)

**Comment by slosar** [comment]
Yes, I was thinking the same, have a list of b(z) and n(z) and just do it at the same time. Note that I'd like to have maybe 10-20 n(z) so think if you can do it in a non-completely brain-dead manner (I think your code now assumes that n<<1 per one FFT pixel, and I think this will remain to be true even in this case).

**Comment by damonge** [comment]
Ok, done. You can now simulate an arbitrary number of populations by concatenating their bias and N(z) files separated by whitespaces. As it is, it will eat up more memory the more populations you add, but this can be fixed if it becomes a problem.

I've tested it a bit and it seems to be running well. Let me know if you see any problems!

**Comment by slosar** [comment]
Fantastic.
How much more memory? The FFT memory x populations or just a bit more?

**Comment by damonge** [comment]
I'm afraid yes. Basically it will allocate an array of ngrid^3 32bit integers for each population, containing the number of sources per cell. This can probably be avoided by adding a bit more code, so let me know.

**Comment by slosar** [comment]
But I don't understand. Last time I looked, you created one delta field and then caclulate the lognormal transformation on the fly -- can't you just do the same with N-species? Also, of course, we do want these populations to be fully correlated... :)

**Comment by damonge** [comment]
I also allocated an array with the number of objects per cell to keep track of which sources I have to allocate and give coordinates and redshifts to in each cell afterwards (this is necessary due to the hybrid parallelization). Now this got multiplied by N_populations. Note that they are still sampled from the same density field.

In any case, forget about all this. I'm just now testing a version of the code that avoids all this extra memory (at the cost of Poisson-sampling twice). I'll push in a couple of hours.

**Comment by damonge** [comment]
Ok, done. Give it a try and let me know if you see any problems!

**Comment by damonge** [comment]
Just pushed a new version of CoLoRe that does this in an optimal way for an arbitrary number of samples defined by their n(z) and b(z). The structure of the config file has changed a bit to make this easier to implement, and to extend to the case where we want to add other tracers (e.g. IM, shear, CMB lensing...)

I won't close this issue. I leave it up to you once you check the new version out.
